### PR TITLE
Add API for accessing expected Pulp actions

### LINKED ISSUE #119: Public interface for getting expected actions on pulp
#### Summary
Ubipop should provide public method for getting expected unassociations and associations that should happen on pulp for ubi repos. Information returned will be used in pub (or other clients) which will run these un/associations and save all received items to push item list.

Input of new method should be content set labels or repository ids.

Output - expected actions on repositories for rpms, srpms, modulemds and modulemd_defaults.

#### Issue Type
- Feature Request

#### Current behavior
There is no way how to get expected pulp actions out of ubipop now.

#### Desired behavior
- ubipop provides public method for getting expected actions on pulp
- ubi won't execute these actions but it'll leave them for caller

#### Use case
Client application needs to know what unit should be un/associated, so caller can execute it.

#### Additional Information
In ubipop there are already methods for calculation of expected pulp actions but it's not exposed publicly and it's implemented inside ubi repos population process.

### PR DESCRIPTION
**Pull request by nathanegillett**
This commit adds a new API which generates a list of the associations
and unassociations expected to be performed by Pulp for the given
configuration.

Fixes #119

### CODE REVIEW

**Comment by nathanegillett** [comment]
Ah, I forgot documentation/example.

**Comment by alexandrevicenzi** [comment]
run tests

**Comment by nathanegillett** [comment]
@rbikar & everyone else, I won't be adding docs for this, as I feel it would require the creation of docs for all of UbiPopulate. So, I think this can be included in the "make public" issue that I've yet to create.

File: ubipop/__init__.py [hunk 1]
```diff
@@ -497,6 +517,28 @@ def _diff_lists_by_attr(self, list_1, list_2, attr):
 
         return diff
 
+    def expected_pulp_actions(self):
+        actions = ExpectedPulpActions([], [])
+
+        Association = namedtuple('Association', ['unit_id', 'type', 'src_repo', 'dest_repo'])
+        Unassociation = namedtuple('Unassociation', ['unit_id', 'type', 'dest_repo'])
+
+        associations, unassociations, _, _ = self._get_pulp_actions(*self._get_current_content())
```
Comment (rbikar) [inline ubipop/__init__.py]: I think there was some misunderstanding with module_defaults units. We still need them to be part of output of this method. 
The issue was with order of execution of association and unassociation of module_defaults units which should be documented.

File: ubipop/__init__.py [hunk 2]
```diff
@@ -132,6 +133,25 @@ def _filter_ubi_conf_list(self, config_list, content_sets, repo_ids):
 
         return filtered_conf_list
 
+    def expected_pulp_actions(self, content_sets=None, repo_ids=None):
```
Comment (rbikar) [inline ubipop/__init__.py]: As this is intended as public API, can you at least add docstring here?

File: ubipop/__init__.py [hunk 3]
```diff
@@ -497,6 +517,28 @@ def _diff_lists_by_attr(self, list_1, list_2, attr):
...
+
+        Association = namedtuple('Association', ['unit_id', 'type', 'src_repo', 'dest_repo'])
+        Unassociation = namedtuple('Unassociation', ['unit_id', 'type', 'dest_repo'])
+
+        associations, unassociations, _, _ = self._get_pulp_actions(*self._get_current_content())
+
+        for item in associations:
+            if item.units:
+                for unit in item.units:
+                    actions.associations.append(Association(unit.name, item.TYPE, item.src_repo.repo_id,
```
Comment (rbikar) [inline ubipop/__init__.py]: unit.name here is not unique identifier of unit that is required for association call to pulp.
e.g. for rpm it's filename, for module it's nsvca. Probably instantiation of Association object with whole unit object would be ok.

File: ubipop/__init__.py [hunk 4]
```diff
@@ -497,6 +517,28 @@ def _diff_lists_by_attr(self, list_1, list_2, attr):
...
+                                                            item.dst_repo.repo_id))
+
+        for item in unassociations:
+            if item.units:
+                for unit in item.units:
+                    actions.unassociations.append(Unassociation(unit.name, item.TYPE,
```
Comment (rbikar) [inline ubipop/__init__.py]: see comment above, the same problem is here.

File: ubipop/__init__.py [hunk 5]
```diff
@@ -497,6 +517,28 @@ def _diff_lists_by_attr(self, list_1, list_2, attr):
 
         return diff
 
+    def expected_pulp_actions(self):
```
Comment (rbikar) [inline ubipop/__init__.py]: The name of this method is the same as in UbiPopulate class. It may be misleading.
Can you consider renaming it?

**Comment by rbikar** [comment]
> @rbikar & everyone else, I won't be adding docs for this, as I feel it would require the creation of docs for all of UbiPopulate. So, I think this can be included in the "make public" issue that I've yet to create.

OK, but please add docstring for public method in UbiPopulate.

**Comment by nathanegillett** [comment]
> Quick question here, why did you abandon #128 PR? Not a big deal but for other reviewers comments there can be valuable.

I agree it wasn't ideal. IIRC, there were problems reverting the work to make the requested changes and syncing this with git/hub.

**Comment by nathanegillett** [comment]
> OK, but please add docstring for public method in UbiPopulate.

Will do! Thanks for the in-line feedback. Very helpful.

**Comment by nathanegillett** [comment]
run tests

File: ubipop/__init__.py [hunk 6]
```diff
@@ -497,10 +539,31 @@ def _diff_lists_by_attr(self, list_1, list_2, attr):
...
 
-    def run_ubi_population(self):
-        current_modules_ft, current_module_defaults_ft, current_rpms_ft, \
-            current_srpms_ft, current_debug_rpms_ft = self._get_current_content()
+    def export_pulp_actions(self):
+        """Formats and returns key data from expected pulp actions for easy external digestion"""
+
+        actions = ExpectedPulpActions([], [])
+
+        Association = namedtuple('Association', ['unit', 'type', 'src_repo', 'dest_repo'])
```
Comment (midnightercz) [inline ubipop/__init__.py]: why did you add new class for Association action when there's already _utils.py AssociateAction?
And if so, why did you create those model classes nested?

Comment (nathanegillett) [reply ubipop/__init__.py]: These un/association tuples are just used to format data from those objects. I suppose I can look into changing those classes to support this use. @rbikar, any objection to that?

Comment (nathanegillett) [reply ubipop/__init__.py]: We can skip these namedtuples, I just think it may be more confusing for the caller of the API. Not too much of a difference, though.

**Comment by nathanegillett** [comment]
@rbikar let me know what you think when you have a chance

**Comment by nathanegillett** [comment]
run tests

**Comment by nathanegillett** [comment]
run tests

**Comment by nathanegillett** [comment]
run tests

**Comment by nathanegillett** [comment]
Sorry, @release-engineering/content-delivery, I've performed a bunch of messy Git operations trying to troubleshoot broken CI which effectively erased the history of this PR. Problem is persisting.

**Comment by nathanegillett** [comment]
Reported broken CI in #134

**Comment by rbikar** [comment]
> @rbikar let me know what you think when you have a chance

@nathanegillett   @midnightercz 

Well I liked that simpler output that was there before.

One thing is that the PulpAction class and its subclasses won't exist in future after _pulp_client is replaced with pulplib, so we will need to do more refactoring here. So I guess creation of namedtuples or something similar should be better.

Second related thing is model classes in _pulp_client won't exists as well, so I guess namedtuple approach or similar will require less refactoring in future (or none if written in generic way) - also this lib will be coupled with Pub pushes so imho less changes -> less number of possible bugs.

**Comment by midnightercz** [comment]
> Well I liked that simpler output that was there before.

You mean like string output? Then you can extend classes with to_str or something similar to get you format that you need. 

> Second related thing is model classes in _pulp_client won't exists as well, so I guess namedtuple approach or similar will require less refactoring in future
Right, but this PR is meant to provide public api for "getting expected actions on pulp". How would you then document that without having publicly documented actions as well?  

_UbiPopulate.expected_pulp_actions return array of some action. Internal structure of actions is unknown._

Isn't better to create simple public class with desired public api for example in models/*.py and once it's in pubtools_pulplib simply remove the model there and only import ModelClass from pubtools_pulplib?
Note that best would be to wait for implementation in pubtools_pulplib

**Comment by rbikar** [comment]
> > Well I liked that simpler output that was there before.
> 
> You mean like string output? Then you can extend classes with to_str or something similar to get you format that you need.

Yeah, that would be even better - some export method in the PulpAction classes.


> 
> > Second related thing is model classes in _pulp_client won't exists as well, so I guess namedtuple approach or similar will require less refactoring in future
> > Right, but this PR is meant to provide public api for "getting expected actions on pulp". How would you then document that without having publicly documented actions as well?
> 
> _UbiPopulate.expected_pulp_actions return array of some action. Internal structure of actions is unknown._
> 
> Isn't better to create simple public class with desired public api for example in models/*.py and once it's in pubtools_pulplib simply remove the model there and only import ModelClass from pubtools_pulplib?
> Note that best would be to wait for implementation in pubtools_pulplib

Personally I would also wait for implementation in pubtools_pulplib so we don't have to do any extra refactoring work here or in Pub.

**Comment by nathanegillett** [comment]
Okay, @rbikar, @midnightercz, we'll hold off on this until pubtools.pulplib is integrated.
